File: back.py
import socket
import sqlite3
import json
import os
from game import game, player
import random
import string


def get_random_string(length):
    letters = string.ascii_lowercase
    return ''.join(random.choice(letters) for i in range(length))

if not os.path.isfile("users.db"):
    file = open("users.db","w+")
    file.close()

conn = sqlite3.connect("users.db")
cursor = conn.cursor()

conn.execute('''CREATE TABLE IF NOT EXISTS users
             (id int,login varchar(256), pass varchar(256), win int, lose int, longest int,temp int)''')
conn.commit()
conn.execute('''CREATE TABLE IF NOT EXISTS games
             (id int,name varchar(256), pass varchar(256),plOneId int,plTwoId int,plOneSnake text,plTwoSnake text,walls text,food text)''')
conn.commit()

sock = socket.socket()
sock.bind(('',8080))
games = {}
print("DATABASE INITED")


def register(player):
    cursor.execute("select COUNT(id) from users where login = '"+player['login']+"'")
    nums = cursor.fetchone()[0]
    if nums==0:
        cursor.execute("select COUNT(id) from users")
        pid = cursor.fetchone()[0]+1
        newp = [pid,player['login'],player['pass'] if player['temp'] == False else get_random_string(16) ,0,0,3,player['temp']]
        print(newp)
        cursor.execute(''' INSERT INTO users VALUES (?,?,?,?,?,?,?) ''',newp)
        conn.commit()
        print(json.dumps({"opStatus":True,"id":pid,"login":player['login']}))
        return json.dumps({"opStatus":True,"id":pid,"login":player['login']})
    else:
        return json.dumps({"opStatus":False})

def checkIfExists(player):
    cursor.execute("select COUNT(id) from users where login = '"+player['login']+"' and pass = '"+player['pass']+"'")
    if cursor.fetchone()[0]==0:
        return json.dumps({"opStatus":False})
    else:
        return json.dumps({"opStatus":True,"id":pid,"login":player['login']})

def exit(player):
    if player["mode"]==3:
        cursor.execute("delete from users where login = '"+player['login']+"' and mode = '"+player['mode']+"' and id = '"+player['id']+"'")
    return json.dumps({"opStatus": True})

def getStat(player):
    cursor.execute("select win,lose,longest from users where login = '"+player['login']+"' and id = '"+player['id']+"' and temp = '"+str(0 if player["mode"]!=3 else 1)+"'")
    res = cursor.fetchone()
    ans = {}
    if res==None:
        ans["opStatus"] = False
    else:
        ans["win"] = res[0]
        ans["lose"] = res[1]
        ans["longest"] = res[2]
    return json.dumps(ans)
   
def getAllGames():      
    cursor.execute("select name,pass from games where plTwoId!='-1'")
    res = cursor.fetchall()
    ans = {}
    ans["opStatus"] = True if res==None else False
    ans["games"] = res
    for elem in ans["games"]:
        elem[1] = (elem[1]!="")
    return json.dumps(ans)

def createGame(game):
    cursor.execute("select COUNT(id) from games where name = '"+game['name']+"'")
    nums = cursor.fetchone()[0]
    ans = {}
    if nums!=0:
        ans["opStatus"] = False
    else:
        ans["opStatus"] = True
        cursor.execute("select COUNT(id) from users")
        gid = cursor.fetchone()[0]+1
        nume = [gid,game['name'],game['pass'],game['pid'],-1,str(""),"",""]
        games[nume] = game.game()
    return ans


def router(cmd: dict):
    if cmd['cmd']=="register":
        rezult = register(cmd["player"])
    elif cmd['cmd']=="exit":
        rezult = exit(cmd["player"])
    elif cmd['cmd']=="check":
        rezult = checkIfExists(cmd["player"])
    elif cmd['cmd']=="getAllInfo":
        rezult = getAllInfo(cmd["player"])
    elif cmd['cmd']=="getAllGames":
        rezult = getAllGames()
    print(type(rezult))
    return rezult

while True:
    sock.listen(1)
    connection, address = sock.accept()
    buf = connection.recv(1024)
    if len(buf) > 0:
        resp = json.loads(str(buf).split("\\r\\n")[-1][:-1])
        answer = ("HTTP/1.1 200 OK\n"
         + "Access-Control-Allow-Origin: *\n"
         + "Content-Type: json\n"
         +router(resp)).encode('utf-8')
        print(answer)
        connection.send(answer)


connection.close()

conn.close()

File: back.rb
#!/usr/bin/env ruby
require 'sinatra'
require 'json'
require 'sinatra/cross_origin'
require 'sqlite3'
require_relative 'game.rb'
File.open("users.db","w") do |file|
end

$db = SQLite3::Database.open 'users.db'


$db.execute "CREATE TABLE IF NOT EXISTS users (id int,login varchar(256), pass varchar(256), win int, lose int, longest int,temp int)"
$db.execute "CREATE TABLE IF NOT EXISTS games (id int,name varchar(256), pass varchar(256),plOneId int,plTwoId int)"

games = {}


def register(player)
  ans = {'opStatus'=>TRUE}
  res = $db.execute 'select count(*) from users where login=?',player['login']
  if res[0][0] == 0
    res = $db.execute 'select count(*) from users'
    id = res[0][0]+1
    $db.execute 'insert into users values(?,?,?,0,0,3,?)', id, player['login'], player['pass'], player['temp']
    ans['pid'] = id
    ans['login'] = player['login']
  else
    ans['opStatus'] = FALSE
  end
   ans.to_json
end

def checkifexists(player)
  ans = {'opStatus'=>true}
  res = $db.execute 'select count(*) from users where login=? and pass=?',player['login'],player['pass']
  if res[0][0]==0
    ans['opStatus'] = FALSE
  else
    ans['pid'] = player['id']
    ans['login'] = player['login']
  end
  ans.to_json
end



def fingame(id)
  games.delete(id)
end

def exitgame(player)
  ans = {'opStatus'=>TRUE}
  res = $db.execute 'select gameId,plOneId,plTwoId from games where plOneId=? or plTwoId=?',player['id'],player['id']
  fingame(res[0][0])
  games['closed']+=1
  $db.execute 'delete from games where gameId = ?',res[0][0]
  ans.to_json
end

def exitapp(player)
  ans = {"opStatus"=>TRUE}
  exitgame(player)
  $db.execute 'delete from users where id=? and login=? and temp=1',res[0][1],res[0][2]
  ans.to_json
end


def getStat(player)
  ans = {"opStatus"=>TRUE}
  res = $db.execute 'select * from users where login=?',player['login']
  if res!=[]
    ans['win'] = res[0][3]
    ans['lose'] = res[0][4]
    ans['longest'] = res[0][5]
  else
    ans['opStatus'] = FALSE
  end
   ans.to_json
end

def getGameStatus()

end

def createGame(game)
  ans = {"opStatus"=>TRUE}
  if $games[game['name']]==nil
    id = $games.lenght()+1
    ans['gameId'] = id
    games[id] = {'game'=>Game.new}
    games[id]['fid'] = game['cid']
  else
    ans['opStatus'] = FALSE
  end
  ans.to_json
end

def logGame(player,game)
  ans = {"opStatus"=>TRUE}
  games[game['id']]['sid'] = player['id']
  ans
end

def getallgames()
  res = $db.execute "select * from games where 'plTwoId'=-1"
  ans = {"opStatus"=>TRUE}
  ans['games'] = res
   ans.to_json
end

def getstatus(player,game)

end

def router(data)
  if data['cmd']=='register'
     register(data['player'])
  elsif data['cmd']=='check'
     checkifexists(data['player'])
  elsif data['cmd']=='getAllInfo'
     getStat(data['player'])
  elsif data['cmd']=='exit'
     exitapp(data['player'])
  elsif data['cmd']=='exitgame'
    exitgame(data['player'])
  elsif data['cmd']=='getAllGames'
     getallgames
  else
     createGame(data['game'])
  end
end



configure do
  enable :cross_origin
end

before do
  response.headers['Access-Control-Allow-Origin'] = '*'
end

options "*" do
  response.headers["Allow"] = "GET, PUT, POST, DELETE, OPTIONS"
  response.headers["Access-Control-Allow-Headers"] = "Authorization, Content-Type, Accept, X-User-Email, X-Auth-Token"
  response.headers["Access-Control-Allow-Origin"] = "*"
  200
end



post "/" do
  data = JSON.parse request.body.read
  puts data
  res = router(data)
  res
end
